docs/4-functional-programming/homework/src/main.ts

```diff
@@ -1,20 +1,51 @@
-import { Either, fromPromise, ap, right, getOrElse, flatten } from './fp/either';
+import { Either, fromPromise, ap, right, flatten, left, getOrElse } from './fp/either';
 import { pipe } from './fp/utils';
 import { fetchClient, fetchExecutor } from './fetching';
 import { ClientUser, ExecutorUser } from './types';
+import { fromNullable, none, getOrElse as ge, Some, isSome } from './fp/maybe';
+import { ordNumber, revert } from './fp/ord';
+import { distance } from './utils';
 
 type Response<R> = Promise<Either<string, R>>
 
 const getExecutor = (): Response<ExecutorUser> => fromPromise(fetchExecutor());
-const getClients = (): Response<Array<ClientUser>> => fromPromise(fetchClient());
+const getClients = (): Response<Array<ClientUser>> => 
+  <Promise<Either<string, ClientUser[]>>>fromPromise(fetchClient().then((clients) => clients.map(({name, demands, position, reward}): ClientUser => (
+   {name, demands: fromNullable(demands), position, reward})
+)))
 
 export enum SortBy {
   distance = 'distance',
   reward = 'reward',
 }
 
 export const show = (sortBy: SortBy) => (clients: Array<ClientUser>) => (executor: ExecutorUser): Either<string, string> => {
-
+  const clientByExecutorDemands = clients.filter(client => {
+    const sortedExecutorPossibilities = executor.possibilities.sort()
+    return client.demands == none || (isSome(client.demands) && client.demands.value.sort().every((element) => sortedExecutorPossibilities.includes(element)))
+  })
+  const clientCount = clientByExecutorDemands.length
+  let possibilitiesMessage
+  if (clientCount === 0){
+    possibilitiesMessage = left('This executor cannot meet the demands of any client!')
+  }
+  else{
+    const sortedClients = clientByExecutorDemands.sort((client1, client2) => {
```

Comment:
It's overall a nice solution, in except for overuse of mutating `Array.prototype.sort` and `if` statements.

Here is another approach of how it can be solved:

```ts
        fold(
            () => left('This executor cannot meet the demands of any client!'),
            (availableClients) => {
                // preparing header string
                const headerString = availableClients.length === clients.length
                    ? 'This executor meets all demands of all clients!'
                    : `This executor meets the demands of only ${availableClients.length} out of ${clients.length} clients`

                // creating function for transforming client into a string about the client
                const stringifyClient = (client: ClientUser) => `\
name: ${client.name}, \
distance: ${distance(executor.position, client.position)}, \
reward: ${client.reward}`

                // creating a functions for sorting
                const byDistanceToExecutor = fromCompare<ClientUser>((client1, client2) => (
                    ordNumber.compare(
                        distance(executor.position, client1.position), distance(executor.position, client2.position)
                    )
                ))
                const byReward = fromCompare<ClientUser>((client1, client2) => ordNumber.compare(client2.reward, client1.reward))

                // with using of our custom pattern matching (something like inline switch case) (can be found in fp/utils.ts)
                // we getting a tuple of [comparator_function(either by distance or by reward), title_of_used_sort_method(either distance or reward)
                const [comparator, sortByTitle] = matcher<SortBy, [Ord<ClientUser>, string]>(
                    [eqStr(SortBy.distance), constant([byDistanceToExecutor, 'distance to executor'])],
                    [eqStr(SortBy.reward), constant([byReward, 'highest reward'])]
                )(sortBy)


                return right(
                    `${headerString}

Available clients sorted by ${sortByTitle}:
                    ${pipe(
                        availableClients,
                        sort(comparator), // sort clients by selected comparator
                        map(stringifyClient) // stringify every client
                    ).join('\n')}`
                )
            }
        )
```